In [7]:
import pandas as pd
ratings_df=pd.read_csv(r"ml-100k\ml-100k\u.data",sep='\t',names=['user_id','item_id','rating_by_user','timestamp'],index_col='user_id')
ratings_df

item_id  rating_by_user  timestamp
user_id                                    
196          242               3  881250949
186          302               3  891717742
22           377               1  878887116
244           51               2  880606923
166          346               1  886397596
...          ...             ...        ...
880          476               3  880175444
716          204               5  879795543
276         1090               1  874795795
13           225               2  882399156
12           203               3  879959583

[100000 rows x 3 columns]

In [9]:
import pandas as pd
user_df=pd.read_csv(r"ml-100k\ml-100k\u.user",sep='|',names=['user_id','Age','Gender','Occupation','user_zipcode'])
user_df

user_id  Age Gender     Occupation user_zipcode
0          1   24      M     technician        85711
1          2   53      F          other        94043
2          3   23      M         writer        32067
3          4   24      M     technician        43537
4          5   33      F          other        15213
..       ...  ...    ...            ...          ...
938      939   26      F        student        33319
939      940   32      M  administrator        02215
940      941   20      M        student        97229
941      942   48      F      librarian        78209
942      943   22      M        student        77841

[943 rows x 5 columns]

In [22]:
import pandas as pd
movies_df=pd.read_csv(r"ml-100k\ml-100k\u.item",sep='|',names=['movie_id','movie_name_year','release_date','imdb_link','unknown',
                    'Action','Adventure','Animation','Children\'s','Comedy','Crime','Documentary','Drama','Fantasy','Film-Noir','Horror','Musical',
                    'Mystery','Romance','Sci-Fi','Thriller','War','Western'], encoding='latin1',index_col=False) #index_col is used so the movie_id values are not considered as index
movies_df

C:\Users\Keert\AppData\Local\Temp\ipykernel_22760\2177103292.py:2: ParserWarning: Length of header or names does not match length of data. This leads to a loss of data with index_col=False.
  movies_df=pd.read_csv(r"ml-100k\ml-100k\u.item",sep='|',names=['movie_id','movie_name_year','release_date','imdb_link','unknown',


movie_id                            movie_name_year release_date  \
0            1                           Toy Story (1995)  01-Jan-1995   
1            2                           GoldenEye (1995)  01-Jan-1995   
2            3                          Four Rooms (1995)  01-Jan-1995   
3            4                          Get Shorty (1995)  01-Jan-1995   
4            5                             Copycat (1995)  01-Jan-1995   
...        ...                                        ...          ...   
1677      1678                          Mat' i syn (1997)  06-Feb-1998   
1678      1679                           B. Monkey (1998)  06-Feb-1998   
1679      1680                       Sliding Doors (1998)  01-Jan-1998   
1680      1681                        You So Crazy (1994)  01-Jan-1994   
1681      1682  Scream of Stone (Schrei aus Stein) (1991)  08-Mar-1996   

      imdb_link                                            unknown  Action  \
0           NaN  http://us.imdb.com/M/title-exact?Toy%20Story%2...       0   
1           NaN  http://us.imdb.com/M/title-exact?GoldenEye%20(...       0   
2           NaN  http://us.imdb.com/M/title-exact?Four%20Rooms%...       0   
3           NaN  http://us.imdb.com/M/title-exact?Get%20Shorty%...       0   
4           NaN  http://us.imdb.com/M/title-exact?Copycat%20(1995)       0   
...         ...                                                ...     ...   
1677        NaN  http://us.imdb.com/M/title-exact?Mat%27+i+syn+...       0   
1678        NaN  http://us.imdb.com/M/title-exact?B%2E+Monkey+(...       0   
1679        NaN      http://us.imdb.com/Title?Sliding+Doors+(1998)       0   
1680        NaN  http://us.imdb.com/M/title-exact?You%20So%20Cr...       0   
1681        NaN  http://us.imdb.com/M/title-exact?Schrei%20aus%...       0   

      Adventure  Animation  Children's  Comedy  ...  Fantasy  Film-Noir  \
0             0          0           1       1  ...        0          0   
1             1          1           0       0  ...        0          0   
2             0          0           0       0  ...        0          0   
3             1          0           0       0  ...        1          0   
4             0          0           0       0  ...        1          0   
...         ...        ...         ...     ...  ...      ...        ...   
1677          0          0           0       0  ...        1          0   
1678          0          0           0       0  ...        0          0   
1679          0          0           0       0  ...        1          0   
1680          0          0           0       0  ...        0          0   
1681          0          0           0       0  ...        1          0   

      Horror  Musical  Mystery  Romance  Sci-Fi  Thriller  War  Western  
0          0        0        0        0       0         0    0        0  
1          0        0        0        0       0         0    1        0  
2          0        0        0        0       0         0    1        0  
3          0        0        0        0       0         0    0        0  
4          0        0        0        0       0         0    1        0  
...      ...      ...      ...      ...     ...       ...  ...      ...  
1677       0        0        0        0       0         0    0        0  
1678       0        0        0        0       1         0    1        0  
1679       0        0        0        0       1         0    0        0  
1680       0        0        0        0       0         0    0        0  
1681       0        0        0        0       0         0    0        0  

[1682 rows x 23 columns]

In [47]:
movies=movies_df.iloc[:,0:2]
movies.head()

movie_id    movie_name_year
0         1   Toy Story (1995)
1         2   GoldenEye (1995)
2         3  Four Rooms (1995)
3         4  Get Shorty (1995)
4         5     Copycat (1995)

In [26]:
ratings_df.shape

(100000, 3)

In [25]:
user_df.shape

(943, 5)

In [48]:
movies.shape

(1682, 2)

# Inferences
So in our datasets, there are 943 users, 1682 to movies and 100000 ratings provided.

## Building the User-Item Matrix

In [29]:
User_Item_Matrix=ratings_df.pivot_table(index=['item_id'],columns=['user_id'],values='rating_by_user')
User_Item_Matrix

user_id  1    2    3    4    5    6    7    8    9    10   ...  934  935  936  \
item_id                                                    ...                  
1        5.0  4.0  NaN  NaN  4.0  4.0  NaN  NaN  NaN  4.0  ...  2.0  3.0  4.0   
2        3.0  NaN  NaN  NaN  3.0  NaN  NaN  NaN  NaN  NaN  ...  4.0  NaN  NaN   
3        4.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  4.0   
4        3.0  NaN  NaN  NaN  NaN  NaN  5.0  NaN  NaN  4.0  ...  5.0  NaN  NaN   
5        3.0  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1678     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1679     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1680     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1681     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   
1682     NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  NaN  ...  NaN  NaN  NaN   

user_id  937  938  939  940  941  942  943  
item_id                                     
1        NaN  4.0  NaN  NaN  5.0  NaN  NaN  
2        NaN  NaN  NaN  NaN  NaN  NaN  5.0  
3        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
4        NaN  NaN  NaN  2.0  NaN  NaN  NaN  
5        NaN  NaN  NaN  NaN  NaN  NaN  NaN  
...      ...  ...  ...  ...  ...  ...  ...  
1678     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1679     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1680     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1681     NaN  NaN  NaN  NaN  NaN  NaN  NaN  
1682     NaN  NaN  NaN  NaN  NaN  NaN  NaN  

[1682 rows x 943 columns]

In [31]:
#Fill the NaN values with 0 as we cannot work with null values and we cannot randomly give ratings using Imputing techniques
User_Item_Matrix.fillna(0,inplace=True)
User_Item_Matrix

user_id  1    2    3    4    5    6    7    8    9    10   ...  934  935  936  \
item_id                                                    ...                  
1        5.0  4.0  0.0  0.0  4.0  4.0  0.0  0.0  0.0  4.0  ...  2.0  3.0  4.0   
2        3.0  0.0  0.0  0.0  3.0  0.0  0.0  0.0  0.0  0.0  ...  4.0  0.0  0.0   
3        4.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  4.0   
4        3.0  0.0  0.0  0.0  0.0  0.0  5.0  0.0  0.0  4.0  ...  5.0  0.0  0.0   
5        3.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
...      ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...  ...   
1678     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1679     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1680     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1681     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   
1682     0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  0.0  ...  0.0  0.0  0.0   

user_id  937  938  939  940  941  942  943  
item_id                                     
1        0.0  4.0  0.0  0.0  5.0  0.0  0.0  
2        0.0  0.0  0.0  0.0  0.0  0.0  5.0  
3        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
4        0.0  0.0  0.0  2.0  0.0  0.0  0.0  
5        0.0  0.0  0.0  0.0  0.0  0.0  0.0  
...      ...  ...  ...  ...  ...  ...  ...  
1678     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1679     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1680     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1681     0.0  0.0  0.0  0.0  0.0  0.0  0.0  
1682     0.0  0.0  0.0  0.0  0.0  0.0  0.0  

[1682 rows x 943 columns]

## Finding the cosine similarity between the items (movies)

In [33]:
from sklearn.metrics import pairwise_distances
import numpy as np

NOTE: When you use pairwise_distances(User_Item_Matrix, metric='cosine'), pairwise_distances treats each row as an individual item (or "sample"), and computes the distance between every pair of rows.

In [36]:
movie_similarity=pairwise_distances(User_Item_Matrix,metric='cosine')
pd.DataFrame(movie_similarity)

0             1             2             3         4         5     \
0     0.000000  5.976178e-01  6.697552e-01  5.450621e-01  0.713286  0.883656   
1     0.597618  1.110223e-16  7.269308e-01  4.974292e-01  0.681164  0.916437   
2     0.669755  7.269308e-01  1.110223e-16  6.751336e-01  0.787043  0.893278   
3     0.545062  4.974292e-01  6.751336e-01  4.440892e-16  0.665761  0.909692   
4     0.713286  6.811638e-01  7.870434e-01  6.657605e-01  0.000000  0.962701   
...        ...           ...           ...           ...       ...       ...   
1677  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  1.000000  1.000000   
1678  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  1.000000  1.000000   
1679  1.000000  1.000000e+00  1.000000e+00  1.000000e+00  1.000000  1.000000   
1680  0.952817  9.217006e-01  1.000000e+00  9.435867e-01  1.000000  1.000000   
1681  0.952817  9.217006e-01  9.031249e-01  9.247823e-01  0.905789  1.000000   

          6         7         8         9     ...      1672  1673      1674  \
0     0.379021  0.518886  0.503712  0.726065  ...  0.964613   1.0  1.000000   
1     0.616597  0.662998  0.744748  0.828918  ...  1.000000   1.0  1.000000   
2     0.627079  0.799206  0.726331  0.841896  ...  1.000000   1.0  1.000000   
3     0.510717  0.509764  0.580956  0.747439  ...  1.000000   1.0  0.905978   
4     0.665231  0.740839  0.727552  0.944547  ...  1.000000   1.0  1.000000   
...        ...       ...       ...       ...  ...       ...   ...       ...   
1677  1.000000  1.000000  1.000000  1.000000  ...  1.000000   1.0  1.000000   
1678  1.000000  1.000000  1.000000  1.000000  ...  1.000000   1.0  1.000000   
1679  1.000000  1.000000  1.000000  1.000000  ...  1.000000   1.0  1.000000   
1680  0.948502  0.917967  0.942640  1.000000  ...  1.000000   1.0  1.000000   
1681  0.948502  1.000000  0.928300  1.000000  ...  1.000000   1.0  1.000000   

          1675      1676  1677  1678  1679      1680      1681  
0     1.000000  0.964613   1.0   1.0   1.0  0.952817  0.952817  
1     1.000000  1.000000   1.0   1.0   1.0  0.921701  0.921701  
2     1.000000  0.967708   1.0   1.0   1.0  1.000000  0.903125  
3     0.905978  0.962391   1.0   1.0   1.0  0.943587  0.924782  
4     1.000000  1.000000   1.0   1.0   1.0  1.000000  0.905789  
...        ...       ...   ...   ...   ...       ...       ...  
1677  1.000000  1.000000   0.0   0.0   0.0  1.000000  1.000000  
1678  1.000000  1.000000   0.0   0.0   0.0  1.000000  1.000000  
1679  1.000000  1.000000   0.0   0.0   0.0  1.000000  1.000000  
1680  1.000000  1.000000   1.0   1.0   1.0  0.000000  1.000000  
1681  1.000000  1.000000   1.0   1.0   1.0  1.000000  0.000000  

[1682 rows x 1682 columns]

In [39]:
np.fill_diagonal(movie_similarity,0)
movie_similarity=pd.DataFrame(movie_similarity)
movie_similarity

0         1         2         3         4         5         6     \
0     0.000000  0.597618  0.669755  0.545062  0.713286  0.883656  0.379021   
1     0.597618  0.000000  0.726931  0.497429  0.681164  0.916437  0.616597   
2     0.669755  0.726931  0.000000  0.675134  0.787043  0.893278  0.627079   
3     0.545062  0.497429  0.675134  0.000000  0.665761  0.909692  0.510717   
4     0.713286  0.681164  0.787043  0.665761  0.000000  0.962701  0.665231   
...        ...       ...       ...       ...       ...       ...       ...   
1677  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   
1678  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   
1679  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000  1.000000   
1680  0.952817  0.921701  1.000000  0.943587  1.000000  1.000000  0.948502   
1681  0.952817  0.921701  0.903125  0.924782  0.905789  1.000000  0.948502   

          7         8         9     ...      1672  1673      1674      1675  \
0     0.518886  0.503712  0.726065  ...  0.964613   1.0  1.000000  1.000000   
1     0.662998  0.744748  0.828918  ...  1.000000   1.0  1.000000  1.000000   
2     0.799206  0.726331  0.841896  ...  1.000000   1.0  1.000000  1.000000   
3     0.509764  0.580956  0.747439  ...  1.000000   1.0  0.905978  0.905978   
4     0.740839  0.727552  0.944547  ...  1.000000   1.0  1.000000  1.000000   
...        ...       ...       ...  ...       ...   ...       ...       ...   
1677  1.000000  1.000000  1.000000  ...  1.000000   1.0  1.000000  1.000000   
1678  1.000000  1.000000  1.000000  ...  1.000000   1.0  1.000000  1.000000   
1679  1.000000  1.000000  1.000000  ...  1.000000   1.0  1.000000  1.000000   
1680  0.917967  0.942640  1.000000  ...  1.000000   1.0  1.000000  1.000000   
1681  1.000000  0.928300  1.000000  ...  1.000000   1.0  1.000000  1.000000   

          1676  1677  1678  1679      1680      1681  
0     0.964613   1.0   1.0   1.0  0.952817  0.952817  
1     1.000000   1.0   1.0   1.0  0.921701  0.921701  
2     0.967708   1.0   1.0   1.0  1.000000  0.903125  
3     0.962391   1.0   1.0   1.0  0.943587  0.924782  
4     1.000000   1.0   1.0   1.0  1.000000  0.905789  
...        ...   ...   ...   ...       ...       ...  
1677  1.000000   0.0   0.0   0.0  1.000000  1.000000  
1678  1.000000   0.0   0.0   0.0  1.000000  1.000000  
1679  1.000000   0.0   0.0   0.0  1.000000  1.000000  
1680  1.000000   1.0   1.0   1.0  0.000000  1.000000  
1681  1.000000   1.0   1.0   1.0  1.000000  0.000000  

[1682 rows x 1682 columns]

## Let's see Recommender Systems

In [53]:
try:
    user_inp = input("Enter the reference movie title based on which the recommendations are to be made: ")
    movie_id = movies[movies['movie_name_year'] == user_inp].index.tolist()
    movie_id = movie_id[0] #extracted the movie_id of the movie given by the user

    movies['Similarity']=movie_similarity.iloc[movie_id]

    print("Recommend movies based on your choice of",user_inp,":\n ",movies.sort_values(['Similarity'],ascending=False)[1:10])
except:
    print("The movie title entered by the user doesnot exist in the list. However, below are the top 10 recommended movies in general")
    print(movies.head(10))

Enter the reference movie title based on which the recommendations are to be made:  Toy Story (1995)


Recommend movies based on your choice of Toy Story (1995) :
        movie_id                           movie_name_year  Similarity
1571      1572            Wend Kuuni (God's Gift) (1982)         1.0
1623      1624                               Hush (1998)         1.0
1624      1625                         Nightwatch (1997)         1.0
1569      1570                    Quartier Mozart (1992)         1.0
1568      1569  Vie est belle, La (Life is Rosey) (1987)         1.0
1567      1568              Vermont Is For Lovers (1992)         1.0
1566      1567                            Careful (1992)         1.0
1565      1566           Man from Down Under, The (1943)         1.0
1564      1565                              Daens (1992)         1.0
